src/core/store/common/tokensByAsset.selectors.ts

```diff
@@ -0,0 +1,86 @@
+import { createSelector } from '@reduxjs/toolkit';
+import { memoize } from 'lodash';
+
+import { toBN } from '@utils';
+import { getConfig } from '@config';
+import { TokenView, Address } from '@types';
+
+import { VaultsSelectors } from '../modules/vaults/vaults.selectors';
+import { LabsSelectors } from '../modules/labs/labs.selectors';
+import { TokensSelectors } from '../modules/tokens/tokens.selectors';
+import { AppSelectors } from '../modules/app/app.selectors';
+import { createToken } from '../modules/tokens/tokens.selectors';
+
+const { selectVaultsMap } = VaultsSelectors;
+const { selectLabsMap } = LabsSelectors;
+const { selectTokensMap, selectTokensUser } = TokensSelectors;
+const { selectServicesEnabled } = AppSelectors;
+
+export const selectDepositTokenOptionsByAsset = createSelector(
+  [selectVaultsMap, selectLabsMap, selectTokensMap, selectTokensUser, selectServicesEnabled],
+  (vaultsMap, labsMap, tokensMap, tokensUser, servicesEnabled) =>
+    memoize((assetAddress?: string): TokenView[] => {
+      if (!assetAddress) return [];
+
+      const { userTokensAddresses, userTokensMap, userTokensAllowancesMap } = tokensUser;
+      const assetData = vaultsMap[assetAddress] ? vaultsMap[assetAddress] : labsMap[assetAddress];
+      if (!assetData) return [];
+
+      const zapperDisabled = !servicesEnabled.zapper && assetData.metadata.zapInWith === 'zapperZapIn';
+      const depositTokenAddresses: Address[] = [];
+      let mainVaultToken: Address;
+      if (zapperDisabled) {
+        mainVaultToken = assetData.token;
+        depositTokenAddresses.push(mainVaultToken);
```

Comment:
This is being done in both if/else blocks, can we move it out

Reply:
the thing is order matter on this one since we want to have the `mainVaultToken` as first element in the array.

Reply:
It's not a problem, you can use `unshift` or just use the spread operator to get it done :)

Reply:
done here https://github.com/yearn/yearn-finance-v3/pull/674/commits/56d21c06d4d0beb30c4cc0cff93519e73646b743 :)